### PR TITLE
Isometry

### PR DESCRIPTION
**Pull request by serraramiro1**

### CODE REVIEW

**Comment by serraramiro1** [comment]
please review when possible @glpuga

File: course/include/isometry.h [hunk 1]
```diff
@@ -72,6 +73,8 @@ class Matrix3 {
   Matrix3 operator*(const Matrix3 &other) const;
   Matrix3 operator*(const Vector3 &other) const;
   Matrix3 operator/(const Matrix3 &other) const;
+  Matrix3 product(const Matrix3 &rhs) const;
+  Vector3 product(const Vector3 &rhs) const;
```
Comment (glpuga) [inline course/include/isometry.h:77]: What are these for? Why not using `operator*()`.

File: course/include/isometry.h [hunk 2]
```diff
@@ -117,4 +120,49 @@ inline std::ostream &operator<<(std::ostream &ss, const Matrix3 &vec) {
...
+
+  Vector3 operator*(const Vector3 &rhs) const {
+    return (rotation_.product(rhs) + translation_);
+  }
+  Isometry operator*(const Isometry &rhs) const {
+    return (Isometry{rotation_.product(rhs.translation_) + translation_,
+                     (rotation_.product(rhs.rotation_))});
+  }
+  Isometry inverse() const {
+    // Only does inverse of translation (do we need inverse of rotation?)
```
Comment (glpuga) [inline course/include/isometry.h]: Yes, you do. An isometry multiplied by its inverse should take you back to the starting point.

File: course/include/isometry.h [hunk 3]
```diff
@@ -117,4 +120,49 @@ inline std::ostream &operator<<(std::ostream &ss, const Matrix3 &vec) {
...
+  Isometry() {
+    rotation_ = Matrix3::kIdentity;
+    translation_ = Vector3::kZero;
+  }
+  Isometry(const Vector3 &trans, const Matrix3 &rot) {
+    translation_ = trans;
+    rotation_ = rot;
+  }
+
+  static const Isometry FromTranslation(const Vector3 &vec);
```
Comment (glpuga) [inline course/include/isometry.h]: Why is this `const`? Does it do anything that the return value is `const` here?

File: course/src/isometry.cc [hunk 4]
```diff
@@ -167,8 +167,61 @@ double Matrix3::det() const {
...
 const Matrix3 Matrix3::kZero = {0., 0., 0., 0., 0., 0., 0., 0., 0.};
 
+const Isometry Isometry::FromTranslation(const Vector3 &vec) {
+  return Isometry{vec, Matrix3::kIdentity};
+}
+
+Isometry Isometry::RotateAround(const Vector3 &direction, const double &value) {
+  Isometry output;
+
+  if (direction == Vector3::kUnitX) {
```
Comment (glpuga) [inline course/src/isometry.cc]: This function should work for **any** `direction` vector. See [this](https://en.wikipedia.org/wiki/Rotation_matrix#Rotation_matrix_from_axis_and_angle). You'll have to make sure you normalize the direction vector.

File: course/test/src/isometry_TEST.cc [hunk 5]
```diff
@@ -16,13 +16,30 @@ namespace ekumen {
...
-//     for (int j = 0; j < 3;j++){
-//       if (abs(a[i][j] - b[i][j]))
-//     }
-//   }
-// } 
+testing::AssertionResult areAlmostEqual(const Matrix3 &a, const Matrix3 &b,
+                                        const double &tolerance) {
+
+  for (int i = 0; i < 3; i++) {
+    for (int j = 0; j < 3; j++) {
```
Comment (glpuga) [inline course/test/src/isometry_TEST.cc]: nit: pre-increment.

File: course/test/src/isometry_TEST.cc [hunk 6]
```diff
@@ -16,13 +16,30 @@ namespace ekumen {
...
+      if ((abs(a[i][j] - b[i][j])) > tolerance) {
+        return testing::AssertionFailure();
+      }
+    }
+  }
+  return testing::AssertionSuccess();
+  }
```
Comment (glpuga) [inline course/test/src/isometry_TEST.cc]: nit: tabulation.

File: course/test/src/isometry_TEST.cc [hunk 7]
```diff
@@ -16,13 +16,30 @@ namespace ekumen {
...
+  
+testing::AssertionResult areAlmostEqual(const Isometry& a, const Isometry& b, const double& tolerance){
+
+  for (int i = 0; i < 3; i++){
```
Comment (glpuga) [inline course/test/src/isometry_TEST.cc]: nit: pre-increment.

**Review by glpuga — CHANGES_REQUESTED** [review]
Very good! Some final comments, and then I'll approve.

File: course/include/isometry.h (hunk 1 above)
Comment (serraramiro1) [reply course/include/isometry.h:77]: The implementation of operator* is not mathematically correct, so had to implement mathematically correct versions of them.

File: course/include/isometry.h (hunk 2 above)
Comment (serraramiro1) [reply course/include/isometry.h]: Done

File: course/include/isometry.h (hunk 3 above)
Comment (serraramiro1) [reply course/include/isometry.h]: Not needed and removed

File: course/src/isometry.cc (hunk 4 above)
Comment (serraramiro1) [reply course/src/isometry.cc]: Done

File: course/test/src/isometry_TEST.cc (hunk 5 above)
Comment (serraramiro1) [reply course/test/src/isometry_TEST.cc]: Done

File: course/test/src/isometry_TEST.cc (hunk 6 above)
Comment (serraramiro1) [reply course/test/src/isometry_TEST.cc]: Addressed

File: course/test/src/isometry_TEST.cc (hunk 7 above)
Comment (serraramiro1) [reply course/test/src/isometry_TEST.cc]: Done

File: course/include/isometry.h (hunk 1 above)
Comment (glpuga) [reply course/include/isometry.h:77]: You're right. I did not remember `operator*` does a scalar multiplication between matrices.

**Review by glpuga — APPROVED** [review]
LGTM, merge away!

**Pull request merged**